### PR TITLE
Upgrade Redis, remove mock-redis, dockerize tests, use SCAN instead of KEYS

### PR DESCRIPTION
**Pull request by UgurGumushan**
Changes
- Upgrades redis version
- Removes redis-mock
- Adjusts mocha to use a dockerized redis instance (needed for removing redis-mock)
(because redis-mock is not compatible with redis client 4+)
- Adjusts CI
- Uses Redis SCAN runction instead of KEYS. Keys is not recommented and blocks IO. 

Backend Adjustments: https://github.com/kontist/backend/pull/13626

Rationale: https://redis.io/commands/keys/ 
`Warning: consider KEYS as a command that should only be used in production environments with extreme care. It may ruin performance when it is executed against large databases. This command is intended for debugging and special operations, such as changing your keyspace layout. Don't use KEYS in your regular application code. If you're looking for a way to find keys in a subset of your keyspace, consider using [SCAN](https://redis.io/commands/scan) or [sets](https://redis.io/topics/data-types#sets).`

Note: When used with the current backend on dev, 1 test fails but it seems to be irrelevant / possibly a flaky test. 

TODO
- [x] Add a test that verifies that getAllPersons() function returns all persons
- [ ] Release an experimental version and test it on backend CI
- [x] Simplify iteration over records (there is an js iterator that is documented)

Resolves: 

- [KRP-1574](https://kontist.atlassian.net/browse/KRP-1574)
- [KRP-1575](https://kontist.atlassian.net/browse/KRP-1575)
- [KRP-1576](https://kontist.atlassian.net/browse/KRP-1576)

[KRP-1576]: https://kontist.atlassian.net/browse/KRP-1576?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

[KRP-1574]: https://kontist.atlassian.net/browse/KRP-1574?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/db.ts [hunk 1]
```diff
@@ -367,25 +383,33 @@ export const findPersonByAccountIBAN = (iban) =>
...
-    .then((keys) => {
-      if (keys.length < 1) {
-        return [];
-      }
-      return redisClient.mgetAsync(keys);
-    })
-    .then((values) => values.map(JSON.parse));
-
-  return webhooks;
+  let cursor = 0;
```
Comment (haipham-kontist) [inline src/db.ts]: :question: are `keys` and `mget`  not supported?

:thinking: the new implementation is much harder to read than the old one. Perhaps if `keys` and `mget` are not supported, can we abstract it?

**Review by haipham-kontist — DISMISSED** [review]
:wrench: can you add some tests to confirm your changes work? The current tests are very limited.

**Comment by haipham-kontist** [comment]
I really like `scanIterator`, look much better now :+1:

**Comment by UgurGumushan** [comment]
@haipham-kontist This is ready for review again

**Comment by UgurGumushan** [comment]
@haipham-kontist is on vacation, so reviews are appreciated: @mbayraktr @lera @p-janik

File: src/db.ts [hunk 2]
```diff
@@ -15,23 +18,27 @@ import {
...
+redisClient
+  .connect()
+  .then(() => {
+    log.info("Redis connection is successful");
+  })
+  .catch((err) => {
+    {
+      log.error(err);
+    }
+  });
```
Comment (haipham-kontist) [inline src/db.ts]: double brackets? :wrench:

File: src/db.ts [hunk 3]
```diff
@@ -302,15 +308,34 @@ export const saveBooking = (accountId, booking) => {
...
-  sort: boolean = false
+/**
+ * Finds persons. When callbackFn is not supplied, loads all persons.
+ * Notes:
+ *  Avoid using this function without a callbackFn
+ *  If you need to find one person, you can use findPerson() instead
+ * @param sort
+ * @param callbackFn
+ */
+export const findPersons = async (
```
Comment (haipham-kontist) [inline src/db.ts:317]: :thinking: I got the error below. (note that `mock-solaris` works, I can login into mobile app)

Screenshot from 2023-06-07 21-28-11

File: src/db.ts [hunk 4]
```diff
@@ -461,46 +479,39 @@ export const saveCardReference = async (cardRef) => {
...
-  const cardData = _(persons)
-    .map((person) => _.get(person, "account.cards", []))
-    .flatten()
-    .value()
-    .find((cd) => cd.card.id === cardId);
-
-  return cardData;
+export const getCardData = async (cardId: string): Promise<Card> => {
+  const personWhoOwnsTheCard = await findPerson(
+    (p) => !!(p?.account?.cards || []).find((cd) => cd?.card?.id === cardId)
```
Comment (haipham-kontist) [inline src/db.ts]: :thought_balloon: you can use `some` instead of `find`.

File: src/routes/postbox.ts [hunk 5]
```diff
@@ -96,12 +94,11 @@ export const listPostboxItems = async (req, res) => {
   res.status(200).send(person.postboxItems || []);
 };
 
-const getPostboxItemById = async (postboxItemId: string) => {
-  const persons = await getAllPersons();
-  const allItems = [].concat(
-    ...persons.map(({ postboxItems }) => postboxItems || [])
+export const getPostboxItemById = async (postboxItemId: string) => {
+  const person = await findPerson(
+    (p) => !!(p.postboxItems ?? []).find((pb) => pb.id === postboxItemId)
```
Comment (haipham-kontist) [inline src/routes/postbox.ts]: :thinking: it would be more readable to replace all `find` by `some` in `findPerson` fn.

File: src/db.ts [hunk 6]
```diff
@@ -461,46 +479,39 @@ export const saveCardReference = async (cardRef) => {
...
-
-  return { person: await getPerson(cardData.card.person_id), cardData };
+  const person = await findPerson((p) => {
+    return !!(p.account?.cards ?? []).find(
+      (c) => !!(c.controls ?? []).find((co) => co.id === id)
+    );
+  });
+  const cardData = (person?.account?.cards ?? []).find(
+    (c) => !!(c.controls ?? []).find((co) => co.id === id)
+  );
```
Comment (haipham-kontist) [inline src/db.ts]: :question: is there a better way to write this? `(c) => !!(c.controls ?? []).find((co) => co.id === id)` was repeated for no reason.

**Comment by haipham-kontist** [comment]
the biggest issue is I can't access mock-solaris in `http://localhost:2091/__BACKOFFICE__` :thinking:. Perhaps I did something wrong?

**Comment by haipham-kontist** [comment]
:warning: you should update the guide in `backend` as well.

File: src/db.ts (hunk 2 above)
Comment (UgurGumushan) [reply src/db.ts]: Related commit: https://github.com/kontist/mock-solaris/pull/114/commits/1822f192ea62292611dc27c31688d03b0bb2250d

File: src/db.ts (hunk 3 above)
Comment (UgurGumushan) [reply src/db.ts:317]: Related commit: https://github.com/kontist/mock-solaris/pull/114/commits/98e05597df1b9c05dde14832ebc3dfe9ab338f13

File: src/db.ts (hunk 4 above)
Comment (UgurGumushan) [reply src/db.ts]: Related commit: https://github.com/kontist/mock-solaris/pull/114/commits/152f423bbbad68c6d888adbf64adfb3c24ed0e74

File: src/routes/postbox.ts (hunk 5 above)
Comment (UgurGumushan) [reply src/routes/postbox.ts]: Related commit: https://github.com/kontist/mock-solaris/pull/114/commits/bc4e4071f154373cfe2a99b139caf81b9af58520

File: src/db.ts (hunk 6 above)
Comment (UgurGumushan) [reply src/db.ts]: This commit makes them non-duplicates: https://github.com/kontist/mock-solaris/pull/114/commits/bc4e4071f154373cfe2a99b139caf81b9af58520

**Comment by UgurGumushan** [comment]
> the biggest issue is I can't access mock-solaris in `http://localhost:2091/__BACKOFFICE__` 🤔. Perhaps I did something wrong?

https://github.com/kontist/mock-solaris/assets/6367201/422201eb-2a0e-40ee-977f-8f81b9601b70

This is me following the new documented flow on readme.md, after recent changes.

Because there is a package change, you may need to run yarn install or docker:start --build

File: tests/db.spec.ts [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+      postboxItems: [mockPostboxItem],
+    };
+    const req = mockReq({ body, headers });
+    const res = mockRes();
+    await createPerson(req, res);
+    const person = await getPostboxItemById(mockPostboxItem.id);
+    expect(person).to.be.ok;
+  });
+
+  it("getPostboxItemById() doesn't throw and returns null if postbox item is not found", async () => {
```
Comment (p-janik) [inline tests/db.spec.ts]: nit some null != undefined, we could replace it descriptions to "falsy value"

Comment (UgurGumushan) [reply tests/db.spec.ts]: I have made this adjustment: https://github.com/kontist/mock-solaris/pull/114/commits/f41947366bab51dc263ce43fee3e252e5887cc9e
(IDE suggests specificity as much as possible)

**Review by p-janik — APPROVED** [review]

**Pull request merged**